src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,395 @@
+    // There are no available nodes to schedule the actor, so just trigger the failed
+    // handler.
+    schedule_failure_handler_(std::move(actor));
+    return;
+  }
+
+  // Update the address of the actor as it is tied to a new node.
+  auto address = actor->GetAddress();
+  address.set_raylet_id(node->node_id());
+  actor->UpdateAddress(address);
+  auto actor_table_data =
+      std::make_shared<rpc::ActorTableData>(actor->GetActorTableData());
+  // The backend storage is reliable in the future, so the status must be ok.
+  RAY_CHECK_OK(actor_info_accessor_.AsyncUpdate(actor->GetActorID(), actor_table_data,
+                                                [this, actor](Status status) {
+                                                  RAY_CHECK_OK(status);
+                                                  // There is no promise that the node the
+                                                  // actor tied to is till alive as the
+                                                  // flush is asynchronously, so just
+                                                  // invoke `Schedule` which will lease
+                                                  // worker directly if the node is still
+                                                  // available or select a new one if not.
+                                                  Schedule(actor);
+                                                }));
+}
+
+std::vector<ActorID> GcsActorScheduler::CancelOnNode(const ClientID &node_id) {
+  // Remove all the actors from the map associated with this node, and return them as they
+  // will be reconstructed later.
+  std::vector<ActorID> actor_ids;
+
+  // Remove all actors in phase of leasing.
+  {
+    auto iter = node_to_actors_when_leasing_.find(node_id);
+    if (iter != node_to_actors_when_leasing_.end()) {
+      actor_ids.insert(actor_ids.end(), iter->second.begin(), iter->second.end());
+      node_to_actors_when_leasing_.erase(iter);
+    }
+  }
+
+  // Remove all actors in phase of creating.
+  {
+    auto iter = node_to_workers_when_creating_.find(node_id);
+    if (iter != node_to_workers_when_creating_.end()) {
+      for (auto &entry : iter->second) {
+        actor_ids.emplace_back(entry.second->TakeAssignedActorID());
+      }
+      node_to_workers_when_creating_.erase(iter);
+    }
+  }
+
+  // Remove the related remote lease client from remote_lease_clients_.
+  remote_lease_clients_.erase(node_id);
+  return actor_ids;
+}
+
+ActorID GcsActorScheduler::CancelOnWorker(const ClientID &node_id,
+                                          const WorkerID &worker_id) {
+  // Remove the worker which is creating actor and return the related actor.
+  return RemoveWorkerInPhaseOfCreating(node_id, worker_id);
+}
+
+void GcsActorScheduler::LeaseWorkerFromNode(std::shared_ptr<GcsActor> actor,
+                                            std::shared_ptr<rpc::GcsNodeInfo> node) {
+  RAY_CHECK(actor && node);
```

Comment:
> I think it is a bit easier to read if you move the AddActorInPhaseOfLeasing call and the check that the node is still alive here (and rename to `LeaseWorkerFromNodeIfStillAlive`). That way, we do not need to make sure to check that the node is still alive every time we call this method.
> 
> Also, I think one way to make the code a bit simpler would be to check whether the actor is still in the leasing map, instead of checking whether the node is still alive. This makes it clear that the invariant is `actor in leasing map <=> try lease request`. Right now, this is hidden because it's hard to see that there is also this invariant that `actor is in leasing map <=> node is still alive`. It's more hash lookups, but I think easier to understand.

Nice!

Reply:
Thanks, I think this is easier to understand now!